### PR TITLE
Created plugin with rule no-deprecated

### PR DESCRIPTION
**Pull request by azhang496**

### CODE REVIEW

**Comment by kaicataldo** [comment]
I think we should be linting our linting rules :P

**Comment by azhang496** [comment]
@kaicataldo so the plugin would have an `.eslintrc` that extends the preset? that's meta

**Comment by kaicataldo** [comment]
Yeah, exactly :) and then lint as part of `npm test`

File: lib/index.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    "no-deprecated": require("./rules/no-deprecated")
+  },
+  configs: {
+    "parserOptions": {
+      "ecmaVersion": 6,
+      "sourceType": "module",
+    },
+    "rules": {
+      "no-deprecated": {
+        "imports": ["nbd/util/extend", "nbd/Promise"],
```
Comment (kaicataldo) [inline lib/index.js]: Shouldn't the actual configuration happen in our preset?

File: lib/rules/no-deprecated.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+          },
+          uniqueItems: true
+        }
+      },
+      additionalProperties: false
+    }]
+  },
+
+  create: function(context) {
+
```
Comment (kaicataldo) [inline lib/rules/no-deprecated.js]: Don't need blank newline

File: lib/rules/no-deprecated.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        node: node,
+        message: '{{item}} is deprecated and should not be used',
+        data: {
+            item: item
+        }
+      });
+    }
+
+    return {
+
```
Comment (kaicataldo) [inline lib/rules/no-deprecated.js]: Don't need blank newline

File: .eslintrc [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
+  "extends": "./node_modules/eslint-preset-behance/.eslintrc",
+  "parserOptions": {
+    "ecmaFeatures": {
+      "jsx": true
```
Comment (kaicataldo) [inline .eslintrc]: What do we need this for?

File: lib/rules/no-deprecated.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        message: '{{item}} is deprecated and should not be used',
+        data: {
+          item
+        }
+      });
+    }
+
+    return {
+      ImportDeclaration(node) {
+        if (blacklist.imports && blacklist.imports.length !== 0) {
```
Comment (kaicataldo) [inline lib/rules/no-deprecated.js]: Can just be `if (blacklist.imports && blacklist.imports.length)`

File: lib/rules/no-deprecated.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+            if (blacklist.imports.indexOf(path) !== -1) {
+              report(node, path);
+            }
+          }
+        }
+      },
+
+      MemberExpression(node) {
+        if (blacklist.expressions && blacklist.expressions.length !== 0) {
```
Comment (kaicataldo) [inline lib/rules/no-deprecated.js]: Can just be `if (blacklist.expressions && blacklist.expressions.length)`

File: .eslintrc (hunk 4 above)
Comment (azhang496) [inline .eslintrc]: yep that should not be there

File: lib/index.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    'no-deprecated': require('./rules/no-deprecated')
+  },
+  configs: {
+    parserOptions: {
+      ecmaVersion: 6,
+      sourceType: 'module',
+    },
+    rules: {
+      'no-deprecated': {
+        imports: ['nbd/util/extend', 'nbd/Promise'],
```
Comment (kaicataldo) [inline lib/index.js]: I think we should leave configuration to the consumer (our preset, in this case)

File: lib/index.js (hunk 1 above)
Comment (azhang496) [inline lib/index.js]: yep was thinking that too, that makes the most sense

File: lib/index.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+'use strict';
+
+// Plugin Definition
+
+module.exports = {
+  rules: {
+    'no-deprecated': require('./rules/no-deprecated')
+  },
+  configs: {
+    parserOptions: {
```
Comment (kaicataldo) [inline lib/index.js]: Actually, do we need any of this configuration? I thought we could just export the rules?

Comment (azhang496) [inline lib/index.js]: nope not necessary, will delete

File: tests/lib/rules/no-deprecated.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    }
+  ],
+
+  invalid: [
+    {
+      code: "import extend from 'nbd/util/extend'",
+      parserOptions: parserOptions,
+      options: [ { imports: ["nbd/util/extend"] } ],
+      errors: [
+        { message: "nbd/util/extend is deprecated and should not be used"}
```
Comment (mikesherov) [inline tests/lib/rules/no-deprecated.js]: please recommend a replacement.

File: tests/lib/rules/no-deprecated.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+      ]
+    },
+    {
+      code: "import Promise from 'nbd/Promise'",
+      parserOptions: parserOptions,
+      options: [ { imports: ["nbd/Promise"] } ],
+      errors: [
+        { message: "nbd/Promise is deprecated and should not be used"}
```
Comment (mikesherov) [inline tests/lib/rules/no-deprecated.js]: please recommend a replacement. Please mention lack of `.finally`

File: tests/lib/rules/no-deprecated.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+      ]
+    },
+    {
+      code: "$.extend(true, {}, fixtures)",
+      parserOptions: parserOptions,
+      options: [ { expressions: ["$.extend"] } ],
+      errors: [
+        { message: "$.extend is deprecated and should not be used"}
```
Comment (mikesherov) [inline tests/lib/rules/no-deprecated.js]: please recommend a replacement.

File: tests/lib/rules/no-deprecated.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+      ]
+    },
+    {
+      code: "$.ajax()",
+      parserOptions: parserOptions,
+      options: [ { expressions: ["$.ajax"] } ],
+      errors: [
+        { message: "$.ajax is deprecated and should not be used"}
```
Comment (mikesherov) [inline tests/lib/rules/no-deprecated.js]: please recommend a replacement.

File: tests/lib/rules/no-deprecated.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    {
+      code: "$.extend(true, {}, fixtures)",
+      parserOptions: parserOptions,
+      options: [ { expressions: ["$.extend"] } ],
+      errors: [
+        { message: "$.extend is deprecated and should not be used"}
+      ]
+    },
+    {
+      code: "$.ajax()",
```
Comment (mikesherov) [inline tests/lib/rules/no-deprecated.js]: single quotes?

**Comment by mikesherov** [comment]
weird that eslint isn't catching the quotes?

File: package.json [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+  "description": "n/a",
+  "keywords": [
+    "eslint",
+    "eslintplugin",
+    "eslint-plugin"
+  ],
+  "author": "Annie Zhang",
+  "main": "lib/index.js",
+  "scripts": {
+    "lint": "./node_modules/.bin/eslint lib/**",
```
Comment (mikesherov) [inline package.json]: ah, you're not linting the tests ;-)

Comment (mikesherov) [inline package.json]: you should be.

File: tests/lib/rules/no-deprecated.js (hunk 10 above)
Comment (azhang496) [inline tests/lib/rules/no-deprecated.js]: what is `.finally`?

**Comment by mikesherov** [comment]
👍

File: lib/rules/no-deprecated.js [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+  create(context) {
+    var blacklist = context.options[0] || {};
+    var imports = blacklist.imports;
+    var expressions = blacklist.expressions;
+
+    function report(node, item, suggestion) {
+      context.report({
+        node,
+        message: '{{item}} is deprecated. use {{suggestion}} instead',
```
Comment (kaicataldo) [inline lib/rules/no-deprecated.js]: Small nit: `{{item}} is deprecated. Use {{suggestion}} instead.`

**Comment by kaicataldo** [comment]
Other than small nit,

**Comment by kaicataldo** [comment]
:+1:

**Comment by hzoo** [comment]
Maybe update the readme with the rule/usage?

**Pull request merged**